File: hu.bme.mit.theta.core/src/main/java/hu/bme/mit/theta/core/type/rattype/RatGtExpr.java
package hu.bme.mit.theta.core.type.rattype;

import static hu.bme.mit.theta.core.type.booltype.BoolExprs.Bool;

import hu.bme.mit.theta.core.Expr;
import hu.bme.mit.theta.core.model.Substitution;
import hu.bme.mit.theta.core.type.abstracttype.GtExpr;
import hu.bme.mit.theta.core.type.booltype.BoolLitExpr;
import hu.bme.mit.theta.core.type.booltype.BoolType;

public final class RatGtExpr extends GtExpr<RatType> {

	private static final int HASH_SEED = 8161;
	private static final String OPERATOR_LABEL = "Gt";

	RatGtExpr(final Expr<RatType> leftOp, final Expr<RatType> rightOp) {
		super(leftOp, rightOp);
	}

	@Override
	public BoolType getType() {
		return Bool();
	}

	@Override
	public BoolLitExpr eval(final Substitution assignment) {
		final RatLitExpr leftOpVal = (RatLitExpr) getLeftOp().eval(assignment);
		final RatLitExpr rightOpVal = (RatLitExpr) getRightOp().eval(assignment);
		return leftOpVal.gt(rightOpVal);
	}

	@Override
	public RatGtExpr with(final Expr<RatType> leftOp, final Expr<RatType> rightOp) {
		if (leftOp == getLeftOp() && rightOp == getRightOp()) {
			return this;
		} else {
			return new RatGtExpr(leftOp, rightOp);
		}
	}

	@Override
	public RatGtExpr withLeftOp(final Expr<RatType> leftOp) {
		return with(leftOp, getRightOp());
	}

	@Override
	public RatGtExpr withRightOp(final Expr<RatType> rightOp) {
		return with(getLeftOp(), rightOp);
	}

	@Override
	public boolean equals(final Object obj) {
		if (this == obj) {
			return true;
		} else if (obj instanceof RatGtExpr) {
			final RatGtExpr that = (RatGtExpr) obj;
			return this.getLeftOp().equals(that.getLeftOp()) && this.getRightOp().equals(that.getRightOp());
		} else {
			return false;
		}
	}

	@Override
	protected int getHashSeed() {
		return HASH_SEED;
	}

	@Override
	protected String getOperatorLabel() {
		return OPERATOR_LABEL;
	}

}

File: subprojects/common/core/src/main/java/hu/bme/mit/theta/core/type/rattype/RatGtExpr.kt
/*
 *  Copyright 2025 Budapest University of Technology and Economics
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package hu.bme.mit.theta.core.type.rattype

import hu.bme.mit.theta.core.model.Valuation
import hu.bme.mit.theta.core.type.Expr
import hu.bme.mit.theta.core.type.abstracttype.GtExpr
import hu.bme.mit.theta.core.type.booltype.BoolLitExpr
import hu.bme.mit.theta.core.type.rattype.RatExprs.Rat
import hu.bme.mit.theta.core.utils.TypeUtils.cast
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
@SerialName("RatGt")
data class RatGtExpr(
    override val leftOp: Expr<RatType>,
    override val rightOp: Expr<RatType>
) : GtExpr<RatType>() {

    companion object {

        @JvmStatic
        fun of(leftOp: Expr<RatType>, rightOp: Expr<RatType>) = RatGtExpr(leftOp, rightOp)

        @JvmStatic
        fun create(leftOp: Expr<*>, rightOp: Expr<*>) = RatGtExpr(cast(leftOp, Rat()), cast(rightOp, Rat()))
    }

    override fun eval(`val`: Valuation): BoolLitExpr {
        val leftOpVal = leftOp.eval(`val`) as RatLitExpr
        val rightOpVal = rightOp.eval(`val`) as RatLitExpr
        return leftOpVal.gt(rightOpVal)
    }

    override fun new(leftOp: Expr<RatType>, rightOp: Expr<RatType>): RatGtExpr =
        of(leftOp, rightOp)

    override fun toString(): String = super.toString()
}

